### ISSUE TITLE
SBT publish failure: Azure Artifacts returns HTTP 203 status code and HTML page on PUT

### ISSUE THREAD

**Issue by 3rwww1** [sbt/sbt#5483]
**EDIT:** my initial assumption (`203 is a success code`) was wrong. From Azure Artifacts perspective,
this is an error. What's "strange" is that changing the PUT UA header to `User-Agent: Apache Maven/3.6.3` actually "tricks" Azure Artifacts into responding with well-formed error/success responses....

In short, the 203 HTTP status code in repsonse of a PUT operation during `sbt publish`  should ~probably~ ~maybe *not*~ **definetely** be considered as an error (actually the publish ~_does_~ **does not** succeed repository side)

The steps to reproduce actually involve setting up Azure Artifacts ~but testing with a proxy returning 203 on PUT should probably do the trick (more than willing to beta test if needed)~

## steps

0. use sbt version ~**1.3.8**~ **1.3.10**

1. Setup an Azure Artifacts repository https://azure.microsoft.com/en-us/services/devops/artifacts/

2. Setup a personal access Token for HTTP auth with Azure DevOps https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=preview-page

3. Setup a `$HOME/.sbt/.credentials` file similar to this

**Note the `realm`  should change depending on the Azure Region/AZ**
```
realm=
host=pkgs.dev.azure.com
user=my_azure_devops_user
password=${ARTIFACTS_PAT}
```

4. Setup `build.sbt` as below

```
name := "MyProject"
organization := "com.myorg"

scalaVersion := "2.11.11"

lazy val sparkVersion = "2.2.0"
lazy val spark = "org.apache.spark"

resolvers += "Spark Packages Repo" at "http://dl.bintray.com/spark-packages/maven"

publishTo := Some("Azure Artifacts" at "https://pkgs.dev.azure.com/myorg/MyProject/_packaging/MyPackages/maven/v1")

dynverVTagPrefix in ThisBuild := false
version in ThisBuild ~= (_.replace('+', '-'))

credentials += Credentials(Path.userHome / ".sbt" / ".credentials")

publishMavenStyle := true

scalacOptions ++= Seq(
  "-feature",
  "-deprecation",
  "-Ywarn-unused-import",
  "-Xfatal-warnings"
)
```
5. run `sbt publish`

## problem

The publish operation fails with the following output:

```
2020-03-25T17:18:34.8773479Z [info] Loading settings for project project-build from plugins.sbt ...
2020-03-25T17:18:35.8097380Z [info] Loading project definition from /home/vsts/work/1/project/project
2020-03-25T17:18:37.2262071Z [info] Loading settings for project project from build.sbt ...
2020-03-25T17:18:37.4750152Z [info] Set current project to MyProject (in build file:/home/vsts/work/1/project/)
2020-03-25T17:18:37.9561066Z [warn] insecure HTTP request is deprecated 'http://dl.bintray.com/spark-packages/maven'; switch to HTTPS or opt-in as ("Spark Packages Repo" at "http://dl.bintray.com/spark-packages/maven").withAllowInsecureProtocol(true)
2020-03-25T17:18:38.1709865Z [warn] insecure HTTP request is deprecated 'http://dl.bintray.com/spark-packages/maven'; switch to HTTPS or opt-in as ("Spark Packages Repo" at "http://dl.bintray.com/spark-packages/maven").withAllowInsecureProtocol(true)
2020-03-25T17:18:38.1722431Z [warn] insecure HTTP request is deprecated 'http://dl.bintray.com/spark-packages/maven'; switch to HTTPS or opt-in as ("Spark Packages Repo" at "http://dl.bintray.com/spark-packages/maven").withAllowInsecureProtocol(true)
2020-03-25T17:18:38.2432721Z [info] Wrote /home/vsts/work/1/project/target/scala-2.11/myproject_2.11-0.5.0-6-383eba0a-20200325-1718.pom
2020-03-25T17:18:41.8256991Z [info] Main Scala API documentation to /home/vsts/work/1/project/target/scala-2.11/api...
2020-03-25T17:18:43.5969913Z [info] Compiling 22 Scala sources to /home/vsts/work/1/project/target/scala-2.11/classes ...
2020-03-25T17:18:50.5521667Z model contains 69 documentable templates
2020-03-25T17:18:53.9704551Z [info] Main Scala API documentation successful.
2020-03-25T17:18:55.9283179Z [info] Done compiling.
2020-03-25T17:18:56.8735362Z [error] java.io.IOException: PUT operation to URL https://pkgs.dev.azure.com/myorg/MyProject/_packaging/MyPackages/maven/v1/com/myorg/myproject_2.11/0.5.0-6-383eba0a-20200325-1718/myproject_2.11-0.5.0-6-383eba0a-20200325-1718.pom failed with status code 203: Non-Authoritative Information
2020-03-25T17:18:56.8739201Z [error] 	at org.apache.ivy.util.url.AbstractURLHandler.validatePutStatusCode(AbstractURLHandler.java:82)
2020-03-25T17:18:56.8744083Z [error] 	at org.apache.ivy.util.url.BasicURLHandler.upload(BasicURLHandler.java:278)
2020-03-25T17:18:56.8747702Z [error] 	at org.apache.ivy.util.url.URLHandlerDispatcher.upload(URLHandlerDispatcher.java:82)
2020-03-25T17:18:56.8750126Z [error] 	at org.apache.ivy.util.FileUtil.copy(FileUtil.java:150)
2020-03-25T17:18:56.8753296Z [error] 	at org.apache.ivy.plugins.repository.url.URLRepository.put(URLRepository.java:84)
2020-03-25T17:18:56.8755780Z [error] 	at sbt.internal.librarymanagement.ConvertResolver$LocalIfFileRepo.put(ConvertResolver.scala:368)
2020-03-25T17:18:56.8758222Z [error] 	at org.apache.ivy.plugins.repository.AbstractRepository.put(AbstractRepository.java:130)
2020-03-25T17:18:56.8761097Z [error] 	at sbt.internal.librarymanagement.ConvertResolver$ChecksumFriendlyURLResolver.put(ConvertResolver.scala:118)
2020-03-25T17:18:56.8764393Z [error] 	at sbt.internal.librarymanagement.ConvertResolver$ChecksumFriendlyURLResolver.put$(ConvertResolver.scala:105)
2020-03-25T17:18:56.8767044Z [error] 	at sbt.internal.librarymanagement.ConvertResolver$$anonfun$defaultConvert$lzycompute$1$PluginCapableResolver$1.put(ConvertResolver.scala:165)
2020-03-25T17:18:56.8769744Z [error] 	at org.apache.ivy.plugins.resolver.RepositoryResolver.publish(RepositoryResolver.java:216)
2020-03-25T17:18:56.8778155Z [error] 	at sbt.internal.librarymanagement.IvyActions$.$anonfun$publish$5(IvyActions.scala:501)
2020-03-25T17:18:56.8783327Z [error] 	at sbt.internal.librarymanagement.IvyActions$.$anonfun$publish$5$adapted(IvyActions.scala:500)
2020-03-25T17:18:56.8786269Z [error] 	at scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:877)
2020-03-25T17:18:56.8788721Z [error] 	at scala.collection.Iterator.foreach(Iterator.scala:941)
2020-03-25T17:18:56.8790948Z [error] 	at scala.collection.Iterator.foreach$(Iterator.scala:941)
2020-03-25T17:18:56.8793158Z [error] 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
2020-03-25T17:18:56.8795473Z [error] 	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
2020-03-25T17:18:56.8798606Z [error] 	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
2020-03-25T17:18:56.8804827Z [error] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
2020-03-25T17:18:56.8810089Z [error] 	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:876)
2020-03-25T17:18:56.8813815Z [error] 	at sbt.internal.librarymanagement.IvyActions$.publish(IvyActions.scala:500)
2020-03-25T17:18:56.8816979Z [error] 	at sbt.internal.librarymanagement.IvyActions$.$anonfun$publish$3(IvyActions.scala:142)
2020-03-25T17:18:56.8820528Z [error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
2020-03-25T17:18:56.8823730Z [error] 	at sbt.internal.librarymanagement.IvyActions$.withChecksums(IvyActions.scala:156)
2020-03-25T17:18:56.8826849Z [error] 	at sbt.internal.librarymanagement.IvyActions$.withChecksums(IvyActions.scala:149)
2020-03-25T17:18:56.8830338Z [error] 	at sbt.internal.librarymanagement.IvyActions$.$anonfun$publish$1(IvyActions.scala:142)
2020-03-25T17:18:56.8833770Z [error] 	at sbt.internal.librarymanagement.IvyActions$.$anonfun$publish$1$adapted(IvyActions.scala:132)
2020-03-25T17:18:56.8837008Z [error] 	at sbt.internal.librarymanagement.IvySbt$Module.$anonfun$withModule$1(Ivy.scala:251)
2020-03-25T17:18:56.8840661Z [error] 	at sbt.internal.librarymanagement.IvySbt.$anonfun$withIvy$1(Ivy.scala:215)
2020-03-25T17:18:56.8843991Z [error] 	at sbt.internal.librarymanagement.IvySbt.sbt$internal$librarymanagement$IvySbt$$action$1(Ivy.scala:77)
2020-03-25T17:18:56.8847186Z [error] 	at sbt.internal.librarymanagement.IvySbt$$anon$1.call(Ivy.scala:87)
2020-03-25T17:18:56.8850573Z [error] 	at xsbt.boot.Locks$GlobalLock.withChannel$1(Locks.scala:95)
2020-03-25T17:18:56.8853980Z [error] 	at xsbt.boot.Locks$GlobalLock.xsbt$boot$Locks$GlobalLock$$withChannelRetries$1(Locks.scala:80)
2020-03-25T17:18:56.8876018Z [error] 	at xsbt.boot.Locks$GlobalLock$$anonfun$withFileLock$1.apply(Locks.scala:99)
2020-03-25T17:18:56.8876506Z [error] 	at xsbt.boot.Using$.withResource(Using.scala:10)
2020-03-25T17:18:56.8876859Z [error] 	at xsbt.boot.Using$.apply(Using.scala:9)
2020-03-25T17:18:56.8877255Z [error] 	at xsbt.boot.Locks$GlobalLock.ignoringDeadlockAvoided(Locks.scala:60)
2020-03-25T17:18:56.8877696Z [error] 	at xsbt.boot.Locks$GlobalLock.withLock(Locks.scala:50)
2020-03-25T17:18:56.8878235Z [error] 	at xsbt.boot.Locks$.apply0(Locks.scala:31)
2020-03-25T17:18:56.8878574Z [error] 	at xsbt.boot.Locks$.apply(Locks.scala:28)
2020-03-25T17:18:56.8878994Z [error] 	at sbt.internal.librarymanagement.IvySbt.withDefaultLogger(Ivy.scala:87)
2020-03-25T17:18:56.8879443Z [error] 	at sbt.internal.librarymanagement.IvySbt.withIvy(Ivy.scala:209)
2020-03-25T17:18:56.8879869Z [error] 	at sbt.internal.librarymanagement.IvySbt.withIvy(Ivy.scala:206)
2020-03-25T17:18:56.8880565Z [error] 	at sbt.internal.librarymanagement.IvySbt$Module.withModule(Ivy.scala:250)
2020-03-25T17:18:56.8881046Z [error] 	at sbt.internal.librarymanagement.IvyActions$.publish(IvyActions.scala:132)
2020-03-25T17:18:56.8881704Z [error] 	at sbt.Classpaths$.$anonfun$publishTask$4(Defaults.scala:2811)
2020-03-25T17:18:56.8882154Z [error] 	at sbt.Classpaths$.$anonfun$publishTask$4$adapted(Defaults.scala:2811)
2020-03-25T17:18:56.8882586Z [error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
2020-03-25T17:18:56.8883060Z [error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
2020-03-25T17:18:56.8883505Z [error] 	at sbt.std.Transform$$anon$4.work(Transform.scala:67)
2020-03-25T17:18:56.8883887Z [error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:281)
2020-03-25T17:18:56.8884327Z [error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:19)
2020-03-25T17:18:56.8884720Z [error] 	at sbt.Execute.work(Execute.scala:290)
2020-03-25T17:18:56.8885072Z [error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:281)
2020-03-25T17:18:56.8885561Z [error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:178)
2020-03-25T17:18:56.8886065Z [error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:37)
2020-03-25T17:18:56.8886489Z [error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2020-03-25T17:18:56.8886951Z [error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
2020-03-25T17:18:56.8887397Z [error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2020-03-25T17:18:56.8887873Z [error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
2020-03-25T17:18:56.8888381Z [error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
2020-03-25T17:18:56.8888797Z [error] 	at java.lang.Thread.run(Thread.java:748)
2020-03-25T17:18:56.8936616Z [error] (publish) java.io.IOException: PUT operation to URL https://pkgs.dev.azure.com/myorg/MyProject/_packaging/MyPackages/maven/v1/com/myorg/myproject_2.11/0.5.0-6-383eba0a-20200325-1718/myproject_2.11-0.5.0-6-383eba0a-20200325-1718.pom failed with status code 203: Non-Authoritative Information
```

## expectation

Publish should succeed AND the server should not respond with a 203

**Comment by mantovani** [comment]
+1 Any idea how to mitigate it ?

I'm having this problem as well.

**Comment by mantovani** [comment]
I'm using this plugin and now is working :)

https://github.com/arktekk/sbt-aether-deploy

**Comment by eed3si9n** [comment]
This is a duplicate of https://github.com/sbt/sbt/issues/2366

**Comment by 3rwww1** [comment]
@eed3si9n This is not the whole story, I'm afraid.
I believe the 203 error was not only caused by #2733.
After a lot of trial/error, it seems linked to **the UA string provided by sbt-ivy,** which is in [BasicURLHandler.java, line 249](https://github.com/sbt/ivy/blob/88d6a93d15f9b029958c1c289a8859e8dfe31a19/src/java/org/apache/ivy/util/url/BasicURLHandler.java#L249)

```java
conn.setRequestProperty("User-Agent", "Apache Ivy/" + Ivy.getIvyVersion());
```
When I replace it with
```java
conn.setRequestProperty("User-Agent", "Apache Maven/3.6.3");
```
I _just works_. I get proper 202, 401, 403, 409 HTTP status codes...
It even works when I simply delete the line :exploding_head:. Repro is quite simple indeed.

This is a "regression" in Azure Artifacts. It actually worked fine (at least in west EU) until mid-march 2020.

Don't know what to think of this :roll_eyes:

**Comment by 3rwww1** [comment]
Full repro case now available:
- modified Ivy repo/branch: https://github.com/3rwww1/ivy/tree/test/azure-artifact-behaves-like-a-2000s-web-app (the interesting lines are here: https://github.com/3rwww1/ivy/commit/0e6e1c64cef7da0930059e9f72fdf568e7516a5f#diff-57f0e4dc47d9fc1270e1504974457cccL251-R252)
- repro case [source code](3rwww1/sbt-test-azure-devops-publishing@5383b9969ae326c3a33689610d3965bbb14022f7), with Azure Pipeline [definition here](https://www.github.com/3rwww1/sbt-test-azure-devops-publishing/blob/5383b9969ae326c3a33689610d3965bbb14022f7/azure-pipelines.yml)
- CI pipeline execution [results](https://dev.azure.com/aironek/sbt-test/_build/results?buildId=24&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=59a85588-b0ba-5043-24c4-d9e29d89c6f6) showing how tinkering with UA made publish pass (lots of error message as I use them for debugging headers)
- PR to add more debug logging to Ivy publish operation: sbt/ivy#38 (still can't enable Ivy's `DEBUG` level through SBT though!)

In short, what I'm doing is:
- building Ivy with modified UA String (now `Apache Maven/3.6.3`)
- building SBT `develop` branch
- try to publish my sample artifact with SBT version `1.4.0-SNAPSHOT`, with `null` credentials and without courier > **it fails with a 203**
- swap `.sbt/boot` Ivy jar with my modified version
- try to publish my sample artifact again now impersonating Maven > **just works**

**Comment by 3rwww1** [comment]
Next steps: PR to allow overriding coursier and Ivy's through an sbt option (`-Dsbt.httpUa="Whatever"`)

**Comment by 3rwww1** [comment]
After some talks with Microsoft support, it appears both Ivy and Maven use `java.net.HttpURLConnection` default accept header, which is `text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2`

Azure Artifacts implement a specific rule for the `Apache Maven/*` UA string that ignore the `Accept` header. However, this rule is not implemented for SBT/Ivy (`Apache Ivy/*` UA string)

I've opened sbt/ivy#40 in order to fix that.

**Comment by 3rwww1** [comment]
@eed3si9n I forgot to PR an update for `librarymanagement`, please find it here : sbt/librarymanagement#347
